uint32 state
uint32 action
